src/Build/Logging/ParallelLogger/ConsoleOutputAligner.cs

```diff
@@ -0,0 +1,143 @@
+    /// </summary>
+    /// <remarks>
+    /// This class is not thread safe.
+    /// </remarks>
+    internal class ConsoleOutputAligner
+    {
+        internal const int ConsoleTabWidth = 8;
+
+        private readonly StringBuilder _reusedStringBuilder = new(1024);
+        private readonly int _bufferWidth;
+        private readonly bool _alignMessages;
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="bufferWidth">Console buffer width. -1 if unknown/unlimited</param>
+        /// <param name="alignMessages">Whether messages are aligned/wrapped into console buffer width</param>
+        public ConsoleOutputAligner(int bufferWidth, bool alignMessages)
+        {
+            _bufferWidth = bufferWidth;
+            _alignMessages = alignMessages;
+        }
+
+        /// <summary>
+        /// Based on bufferWidth split message into multiple lines and indent if needed.
+        /// TAB character are interpreted by standard Console logic. 
+        /// </summary>
+        /// <param name="message">Input message. May contains tabs and new lines. Both \r\n and \n is supported but replaced into current environment new line.</param>
+        /// <param name="prefixAlreadyWritten">true if message already contains prefix (message context, timestamp, etc...).</param>
+        /// <param name="prefixWidth">Width of the prefix. Every line in result string will be indented by this number of spaces except 1st line with already written prefix.</param>
+        /// <returns>Aligned message ready to be written to Console</returns>
+        /// <remarks>
+        /// For optimization purposes this method uses single <see cref="StringBuilder"/> instance. This makes this method non thread safe.
+        /// Calling side is expected this usage is non-concurrent. This shall nto be an issue as it is expected that writing into Console shall be serialized anyway.
+        /// </remarks>
+        public string AlignConsoleOutput(string message, bool prefixAlreadyWritten, int prefixWidth)
+        {
+            int i = 0;
+            int j = message.IndexOfAny(MSBuildConstants.CrLf);
+
+            StringBuilder sb = _reusedStringBuilder;
+            // prepare reused StringBuilder instance for new use.
+            sb.Length = 0;
+            // The string contains new lines, treat each new line as a different string to format and send to the console
+            while (j >= 0)
+            {
+                AlignAndIndentLineOfMessage(sb, prefixAlreadyWritten, prefixWidth, message, i, j - i);
+                i = j + (message[j] == '\r' && (j + 1) < message.Length && message[j + 1] == '\n' ? 2 : 1);
+                j = i < message.Length ? message.IndexOfAny(MSBuildConstants.CrLf, i) : -1;
+            }
+
+            // Process rest of message
+            AlignAndIndentLineOfMessage(sb, prefixAlreadyWritten, prefixWidth, message, i, message.Length - i);
+
+            return sb.ToString();
```

Comment:
if sb grew hugely during this work, say, it's now 50K long, should you release it and replace it with a fresh 1024 char one?

Reply:
I though about it and I decided to keep the instance big. With current code instance of `ConsoleOutputAligner` is recycled after end of build. So that huge `StringBuilder` instance will have lifetime of build only.

I do not see much benefit from recycling SB during build as there is possibility we will need that big SB frequently.
I am not too afraid about OOM regression, as changes in this PR makes memory traffic along with working set considerably smaller.